### PR TITLE
AMDGPU: add hotswap entry trampoline core

### PR DESCRIPTION
**Pull request by harsh-amd**
Stack:
1. This PR - core hotswap entry trampoline implementation, explicit COMGR options API, unit coverage, and minimal public LIT coverage.
2. #3098 - expanded LIT coverage layered on this PR.

The temporary ROCm base branch users/harsh/comgr-hotswap-entry-core mirrors this PR head so #3098 shows only the incremental LIT diff.

Summary:
- Adds opt-in gfx125x hotswap entry trampoline generation and descriptor fixups.
- Exposes `amd_comgr_hotswap_rewrite_with_options` and `AMD_COMGR_HOTSWAP_REWRITE_FLAG_ENTRY_TRAMPOLINES` so callers request entry trampolines explicitly per rewrite call.
- Removes COMGR production dependence on `AMD_COMGR_HOTSWAP_ENTRY_TRAMPOLINES`; COMGR no longer uses that env var as the integration contract.
- Keeps the existing `amd_comgr_hotswap_rewrite` ABI-compatible default path without entry trampolines.
- Wires the rewrite pipeline to fail closed when requested entry trampolines cannot be installed.
- Tightens entry-stub idempotency by validating decoded operand shape and that the stub targets earlier .text.
- Aligns appended entry stubs by virtual address and preserves the instruction-prefetch guard.
- Makes program-header growth overflow handling fail closed instead of skipping malformed headers.
- Adds ELF helpers for enumerating/updating kernel descriptors and SGPR metadata.
- Adds focused unit coverage for the ELF helpers, trampoline builder, descriptor rewrites, SGPR metadata updates, idempotency, operand validation, alignment, and failure paths.
- Updates the shared `hotswap-rewrite` LIT driver to exercise `amd_comgr_hotswap_rewrite_with_options`.
- Adds minimal public LIT coverage for default-off behavior, explicit options-API opt-in entry trampolines, invalid options, and fail-closed scratch-SGPR exhaustion.

Verification:
- Branch head: 80cbb2836313.
- Stack top head: f80a756987f1.
- `python3 amd/comgr/utils/check_api_consistency.py --comgr-dir amd/comgr` passed.
- `cmake --build build --target hotswap-rewrite -- -j8` passed.
- `build/bin/llvm-lit -sv build/tools/comgr/test-lit --filter 'hotswap-rewrite.c|hotswap-kernel-entry-trampoline.s'` passed: 2 tests on #3008.
- `build/bin/llvm-lit -sv build/tools/comgr/test-lit --filter hotswap` passed: 53 tests on the stack top.
- ASAN `cmake --build build-asan --target hotswap-rewrite -- -j8` passed.
- ASAN `build-asan/bin/llvm-lit -sv build-asan/tools/comgr/test-lit --filter hotswap` passed: 53 tests on the stack top.
- Earlier full normal `check-comgr` passed after the options API update: lit 72 passed/11 unsupported; unit tests passed; CTest 31/31 passed.
- Earlier full ASAN `check-comgr` passed after the options API update: lit 72 passed/11 unsupported; unit tests passed; CTest 31/31 passed.

### CODE REVIEW

File: amd/comgr/test-unit/HotswapMCTest.cpp [hunk 1]
```diff
@@ -681,8 +772,12 @@ TEST(AddTid, StoreTrampolineThroughBuildTrampoline) {
...
-  EXPECT_EQ(Decoded[0].Mnemonic, "v_mbcnt_lo_u32_b32");
-  EXPECT_EQ(Decoded[5].Mnemonic, "ds_store_b32");
-  EXPECT_EQ(Decoded[6].Mnemonic, "s_branch");
+  const llvm::StringRef Expected[] = {"v_mbcnt_lo_u32_b32",
+                                      "v_mbcnt_hi_u32_b32",
+                                      "v_lshlrev_b32",
+                                      "v_add_nc_u32",
+                                      "v_and_b32",
+                                      "ds_store_b32",
+                                      "s_branch"};
```
Comment (chinmaydd) [inline amd/comgr/test-unit/HotswapMCTest.cpp:916]: This is an improvement, but it feels rather incomplete if we dont check the operands of the rewritten instructions.

We should definitely have integration tests, but IMO it should also be fixed in Unit tests.

Comment (chinmaydd) [reply amd/comgr/test-unit/HotswapMCTest.cpp:916]: Same for others...

File: amd/comgr/src/hotswap/README.md [hunk 2]
```diff
@@ -6,9 +6,24 @@ target ISA names, then returns a new executable code object with the applicable
...
+path for heavier cross-ISA transformations. The same-family stepping patches and
+optional entry trampolines are implemented in the surrounding COMGR source files
+and are exposed through `amd_comgr_hotswap_rewrite`.
+
+## Supported transformations
+
+| Transformation | Status |
+| -------------- | ------ |
+| gfx1250 B0 to A0 | Supported |
+| gfx12.5/gfx125 entry trampolines | Supported, opt-in |
```
Comment (chinmaydd) [inline amd/comgr/src/hotswap/README.md]: Should this be 125x or 125 ? Maybe there's a naming convention update I'm unaware of.

File: amd/comgr/test-lit/hotswap-kernel-entry-trampoline.s [hunk 3]
```diff
@@ -0,0 +1,72 @@
+// COM: HotSwap redirects kernel descriptors to appended PC-relative entry
+// COM: stubs only when the entry-trampoline flag is enabled.
+
+// RUN: %clang -target amdgcn-amd-amdhsa -mcpu=gfx1250 -nostdlib %s -o %t.elf
+
+// RUN: AMD_COMGR_HOTSWAP_ENTRY_TRAMPOLINES=1 hotswap-rewrite %t.elf \
+// RUN:   amdgcn-amd-amdhsa--gfx1250 amdgcn-amd-amdhsa--gfx1250 \
```
Comment (chinmaydd) [inline amd/comgr/test-lit/hotswap-kernel-entry-trampoline.s:7]: This style of the triple is going to be deprecated soon. But I suppose we will take of it after that change flows in Comgr from upstream...

**Comment by chinmaydd** [comment]
I have read through the code generally, but admittedly, there's a lot of it. I'm sure to have missed things. I ran it through Claude and here are some issues that we should probably take a look at:

---

1. **`patchDebugSections` gets `Deferred` but the ELF grew by `Growth`** (`comgr-hotswap-b0a0.cpp:636`): symbols/line-rows cover only B0→A0 trampolines while the address shift includes entry-stub bytes; in the entry-only path zero symbols are added for a non-zero shift (latent: `patchDebug*` are weak no-ops today).
2. **`isKernelEntryTrampoline` matches opcodes only, not the jump target** (`comgr-hotswap-entry-trampoline.cpp:144`/`318`): a real kernel beginning with the same 6-instruction prologue is misread as "already stubbed" and silently skipped.
3. **Stub 256-byte alignment computed on file offset, not vaddr** (`comgr-hotswap-entry-trampoline.cpp:343`): `textSize`-relative rounding misaligns the redirected entry whenever `textAddr % 256 != 0`.
4. **`getKernelSgprCount` drops the KD fallback for malformed/partial metadata** (`comgr-hotswap-elf.cpp:691`): any present-but-bad AMDGPU note now hard-returns `nullopt`, aborting the rewrite where the old descriptor fallback succeeded.
5. **`updateKernelDescriptorSgprCount` writes RSRC1 from an `8` baseline while metadata is the true source** (`comgr-hotswap-elf.cpp:648`): leaves descriptor/metadata SGPR counts divergent on GFX12 and can over-reserve up to a 112-SGPR granule.
6. **`adjustProgramHeaders` swallows overflow with `continue` instead of `return false`** (`comgr-hotswap-elf.cpp:982`): a segment can be left un-grown, yielding a "successful" rewrite whose `PT_LOAD` no longer spans the enlarged `.text`.
7. **`adjustSymbolValues` relocates symbols by file-offset order, not vaddr** (`comgr-hotswap-elf.cpp:1135`): a section whose file order differs from vaddr order gets its address shifted but its symbols' `st_value` left stale.
8. **Duplicate base-name `.kd` symbols resolve inconsistently across the two ElfViews** (`comgr-hotswap-entry-trampoline.cpp`): one descriptor can be rewritten twice (last-writer-wins) while the other keeps an orphaned stub.
9. **`descriptorAlreadyTargetsEntryStub` treats an `entryVAddr()` failure as "not a stub"** (`comgr-hotswap-entry-trampoline.cpp:318`): a descriptor at the overflow boundary is re-processed and gets a second stub, defeating idempotency.

---

AFAICT, these are really corner cases that we probably wont encounter in well-formed kernels. In that sense, I think we should be okay.

File: amd/comgr/test-lit/hotswap-kernel-entry-trampoline.s (hunk 3 above)
Comment (harsh-amd) [reply amd/comgr/test-lit/hotswap-kernel-entry-trampoline.s:7]: `This style of the triple is going to be deprecated soon.`
This is news to me. What is it going to be moving forward?

`But I suppose we will take of it after that change flows in Comgr from upstream...`
Thank you :)

File: amd/comgr/src/hotswap/README.md (hunk 2 above)
Comment (harsh-amd) [reply amd/comgr/src/hotswap/README.md]: This should be `gfx125x`.I updated the README wording, the public error text, the PR description, and the stacked lit coverage naming/comment to use `gfx125x`. That matches the naming used by the related ROCm Systems change for the loader-side trampoline removal.

File: amd/comgr/test-unit/HotswapMCTest.cpp (hunk 1 above)
Comment (harsh-amd) [reply amd/comgr/test-unit/HotswapMCTest.cpp:916]: Fixed. Unit tests now compare decoded `MCInst` operands against expected assembled instructions for ADDTID load/store bodies and the entry trampoline.

**Comment by harsh-amd** [comment]
> I have read through the code generally, but admittedly, there's a lot of it. I'm sure to have missed things. I ran it through Claude and here are some issues that we should probably take a look at:
> 
> 1. **`patchDebugSections` gets `Deferred` but the ELF grew by `Growth`** (`comgr-hotswap-b0a0.cpp:636`): symbols/line-rows cover only B0→A0 trampolines while the address shift includes entry-stub bytes; in the entry-only path zero symbols are added for a non-zero shift (latent: `patchDebug*` are weak no-ops today).
> 2. **`isKernelEntryTrampoline` matches opcodes only, not the jump target** (`comgr-hotswap-entry-trampoline.cpp:144`/`318`): a real kernel beginning with the same 6-instruction prologue is misread as "already stubbed" and silently skipped.
> 3. **Stub 256-byte alignment computed on file offset, not vaddr** (`comgr-hotswap-entry-trampoline.cpp:343`): `textSize`-relative rounding misaligns the redirected entry whenever `textAddr % 256 != 0`.
> 4. **`getKernelSgprCount` drops the KD fallback for malformed/partial metadata** (`comgr-hotswap-elf.cpp:691`): any present-but-bad AMDGPU note now hard-returns `nullopt`, aborting the rewrite where the old descriptor fallback succeeded.
> 5. **`updateKernelDescriptorSgprCount` writes RSRC1 from an `8` baseline while metadata is the true source** (`comgr-hotswap-elf.cpp:648`): leaves descriptor/metadata SGPR counts divergent on GFX12 and can over-reserve up to a 112-SGPR granule.
> 6. **`adjustProgramHeaders` swallows overflow with `continue` instead of `return false`** (`comgr-hotswap-elf.cpp:982`): a segment can be left un-grown, yielding a "successful" rewrite whose `PT_LOAD` no longer spans the enlarged `.text`.
> 7. **`adjustSymbolValues` relocates symbols by file-offset order, not vaddr** (`comgr-hotswap-elf.cpp:1135`): a section whose file order differs from vaddr order gets its address shifted but its symbols' `st_value` left stale.
> 8. **Duplicate base-name `.kd` symbols resolve inconsistently across the two ElfViews** (`comgr-hotswap-entry-trampoline.cpp`): one descriptor can be rewritten twice (last-writer-wins) while the other keeps an orphaned stub.
> 9. **`descriptorAlreadyTargetsEntryStub` treats an `entryVAddr()` failure as "not a stub"** (`comgr-hotswap-entry-trampoline.cpp:318`): a descriptor at the overflow boundary is re-processed and gets a second stub, defeating idempotency.
> 
> AFAICT, these are really corner cases that we probably wont encounter in well-formed kernels. In that sense, I think we should be okay.

Thanks. I fixed the entry-trampoline issues from this list: opcode-only idempotency, vaddr alignment, program-header overflow, entryVAddr idempotency failure, and missing operand checks in unit tests.

 I left the debug/metadata/symbol-order/duplicate-name items unchanged since they are broader ELF/debug/metadata policy issues outside this PR’s entry-trampoline path.

File: amd/comgr/test-lit/hotswap-kernel-entry-trampoline.s (hunk 3 above)
Comment (chinmaydd) [reply amd/comgr/test-lit/hotswap-kernel-entry-trampoline.s:7]: You can refer to the PR stack here to get a clue : https://github.com/llvm/llvm-project/pull/206488/changes.

I am trying to get a hold of the presentation that Matt gave to the compiler team yesterday. I'll share it with others when I have it.

Effectively, moving towards:

`amdgpu12.50-amd-amdhsa`

**Comment by harsh-amd** [comment]
Also I modified this PR so comgr no longer depends on the global environment var. The idea is that rocr runtime will call comgr with the appropriate rewrite options depending on the value of that global environment var. The existing amd_comgr_hotswap_rewrite path remains ABI-compatible and keeps entry trampolines off by default, while callers like rocm-systems can opt in only for the gfx12.5 cases they own using amd_comgr_hotswap_rewrite_with_options.

**Comment by chinmaydd** [comment]
> Also I modified this PR so comgr no longer depends on the global environment var. The idea is that rocr runtime will call comgr with the appropriate rewrite options depending on the value of that global environment var

I see, thanks. Yes, I think thats the right thing to do.

However, the LIT test (and its driver) does not exercise the new path yet.

**Comment by harsh-amd** [comment]
Done. I moved the minimal public LIT coverage into this PR so #3008 now exercises the new options API directly. The shared `hotswap-rewrite` driver has `--entry-trampolines` and invalid-options modes, and `hotswap-kernel-entry-trampoline.s` now uses `amd_comgr_hotswap_rewrite_with_options` through that driver instead of the removed env var path. #3098 was rebased so it only carries the expanded LIT matrix.

Verified normal and ASAN hotswap lit: 53 tests passed in each build.

File: amd/comgr/src/comgr-hotswap-b0a0.cpp [hunk 4]
```diff
@@ -520,79 +522,125 @@ static void runScratchVerification(WritableMemoryBuffer &OutBuf,
...
+    if (!Result) {
+      log() << "hotswap: error: retargetCodeObject: "
+            << "getNewUninitMemBuffer(" << ElfSize
+            << ") failed (out of memory) for the no-op output copy.\n";
+      return AMD_COMGR_STATUS_ERROR_OUT_OF_RESOURCES;
+    }
+    std::memcpy(Result->getBufferStart(), ElfData, ElfSize);
+    Out = std::move(Result);
+    return AMD_COMGR_STATUS_SUCCESS;
+  }
```
Comment (chinmaydd) [inline amd/comgr/src/comgr-hotswap-b0a0.cpp:548]: I think we're almost there.

This no-op behavior should be documented somewhere

Comment (chinmaydd) [reply amd/comgr/src/comgr-hotswap-b0a0.cpp:548]: Its also weird that there's no way for the caller to know whether this was exercised.

File: amd/comgr/test-lit/comgr-sources/hotswap-rewrite.c [hunk 5]
```diff
@@ -8,14 +8,43 @@
 ///
 /// \file
 /// Canonical hotswap input/output driver for lit tests. Loads an ELF, runs
-/// amd_comgr_hotswap_rewrite, and optionally dumps the output and/or checks
-/// that a second rewrite produces identical output (idempotency).
+/// the hotswap rewrite API, and optionally dumps the output and/or checks that
+/// a second rewrite produces identical output (idempotency).
```
Comment (chinmaydd) [inline amd/comgr/test-lit/comgr-sources/hotswap-rewrite.c]: Do we need to claim idempotency? This is a strong claim and I wonder if we have tests for it.

When B0A0 and entry trampolines run together, the B0-to-A0 pass re-decodes the whole grown `.text` (now including the appended stub pool) and has no 'skip existing stub' guard, unlike the entry pass (Work-empty check).

**Comment by chinmaydd** [comment]
There's also a bunch of redundancy in implementation, but we can take that up in the refactor.

File: amd/comgr/src/comgr-hotswap-b0a0.cpp (hunk 4 above)
Comment (harsh-amd) [reply amd/comgr/src/comgr-hotswap-b0a0.cpp:548]: I agree this is not ideal, but I would prefer not to grow the API shape in this PR. For now, callers that need to know can compare input/output bytes. If ROCr needs structured pass-result reporting, we should add that as a follow-up API extension.

Comment (harsh-amd) [reply amd/comgr/src/comgr-hotswap-b0a0.cpp:548]: Done. I documented this in the public hotswap API comments and the HotSwap README. SUCCESS means comgr produced a valid output code object, not that bytes necessarily changed; if source/target/options select no enabled rewrite, output is a copy of the input.

File: amd/comgr/test-lit/comgr-sources/hotswap-rewrite.c (hunk 5 above)
Comment (harsh-amd) [reply amd/comgr/test-lit/comgr-sources/hotswap-rewrite.c]: Agreed. I reworded the driver comment. The intent is only that the test driver can rerun the same request and compare outputs, not to claim global idempotency for every combination of rewrite options.

**Comment by harsh-amd** [comment]
Sounds good. I will leave the broader cleanup for the follow-up refactor and keep this PR focused on the API and entry-trampoline behavior.

**Comment by chinmaydd** [comment]
Thanks. This mostly LGTM.

@lamb-j is planning to look as well, so I'd recommend we wait for his inputs.

**Review by chinmaydd — APPROVED** [review]

**Comment by harsh-amd** [comment]
> Thanks. This mostly LGTM.
> 
> @lamb-j is planning to look as well, so I'd recommend we wait for his inputs.

Sounds good. Can we land this and address his comments in the follow up PR?

**Comment by chinmaydd** [comment]
Sure. When I had a chat with him, I think a concern was about adding a new Comgr API without incrementing the semantic versioning. Although, in this case I dont think that would be an issue.

Its okay to land for now.

**Comment by lamb-j** [comment]
Filed https://github.com/ROCm/llvm-project/pull/3134

Ideally in the future, we do the version bump in the same commit as the API introduction. It's easier to cherry-pick, revert, etc. if they're together.

Also if we're really unlucky, a branch can be cut between the API introduction and the version bump, which exposes us to some pretty big inconsistencies. Not as big a deal for minor bumps compared to major, but still not ideal
